File: generated-code/kotlin/codecraft/FileReadWrite/src/main/kotlin/trans_gen_test/model/AttackAction.kt
package trans_gen_test.model

import trans_gen_test.util.StreamUtil

/**
 * Attack action
 */
class AttackAction {
    /**
     * If specified, target entity's ID
     */
    var target: Int?
    /**
     * If specified, configures auto attacking
     */
    var autoAttack: trans_gen_test.model.AutoAttack?

    constructor(target: Int?, autoAttack: trans_gen_test.model.AutoAttack?) {
        this.target = target
        this.autoAttack = autoAttack
    }

    /**
     * Write AttackAction to output stream
     */
    @Throws(java.io.IOException::class)
    fun writeTo(stream: java.io.OutputStream) {
        val targetValue = target
        if (targetValue == null) {
            StreamUtil.writeBoolean(stream, false)
        } else {
            StreamUtil.writeBoolean(stream, true)
            StreamUtil.writeInt(stream, targetValue)
        }
        val autoAttackValue = autoAttack
        if (autoAttackValue == null) {
            StreamUtil.writeBoolean(stream, false)
        } else {
            StreamUtil.writeBoolean(stream, true)
            autoAttackValue.writeTo(stream)
        }
    }

    /**
     * Get string representation of AttackAction
     */
    override fun toString(): String {
        var stringBuilder = StringBuilder("AttackAction { ")
        stringBuilder.append("target: ")
        stringBuilder.append(target)
        stringBuilder.append(", ")
        stringBuilder.append("autoAttack: ")
        stringBuilder.append(autoAttack)
        stringBuilder.append(" }")
        return stringBuilder.toString()
    }

    companion object {
        /**
         * Read AttackAction from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): AttackAction {
            var target: Int?
            if (StreamUtil.readBoolean(stream)) {
                target = StreamUtil.readInt(stream)
            } else {
                target = null
            }
            var autoAttack: trans_gen_test.model.AutoAttack?
            if (StreamUtil.readBoolean(stream)) {
                autoAttack = trans_gen_test.model.AutoAttack.readFrom(stream)
            } else {
                autoAttack = null
            }
            return AttackAction(target, autoAttack)
        }
    }
}

File: generated-code/scala/codecraft/FileReadWrite/src/main/scala/trans_gen_test/model/AttackAction.scala
package trans_gen_test.model

import trans_gen_test.util.StreamUtil

/**
 * Attack action
 *
 * @param target If specified, target entity's ID
 * @param autoAttack If specified, configures auto attacking
 */
case class AttackAction(target: Option[Int], autoAttack: Option[trans_gen_test.model.AutoAttack]) {
    /**
     * Write AttackAction to output stream
     */
    def writeTo(stream: java.io.OutputStream) {
        target match {
            case None => StreamUtil.writeBoolean(stream, false)
            case Some(value) => {
                StreamUtil.writeBoolean(stream, true)
                StreamUtil.writeInt(stream, value)
            }
        }
        autoAttack match {
            case None => StreamUtil.writeBoolean(stream, false)
            case Some(value) => {
                StreamUtil.writeBoolean(stream, true)
                value.writeTo(stream)
            }
        }
    }

    /**
     * Get string representation of AttackAction
     */
    override def toString(): String = {
        var stringBuilder = new StringBuilder("AttackAction { ")
        stringBuilder.append("target: ")
        stringBuilder.append(target)
        stringBuilder.append(", ")
        stringBuilder.append("autoAttack: ")
        stringBuilder.append(autoAttack)
        stringBuilder.append(" }")
        stringBuilder.toString()
    }
}

object AttackAction {
    /**
     * Read AttackAction from input stream
     */
    def readFrom(stream: java.io.InputStream): AttackAction = AttackAction(
        if (StreamUtil.readBoolean(stream)) Some(
            StreamUtil.readInt(stream)
        ) else None,
        if (StreamUtil.readBoolean(stream)) Some(
            trans_gen_test.model.AutoAttack.readFrom(stream)
        ) else None
    )
}